File: openrgb-wrapper/src/main/java/de/lars/openrgbwrapper/models/Mode.java
package de.lars.openrgbwrapper.models;

public class Mode {

    private String name;
    private int value;
    private int flags;
    private int speedMin;
    private int speedMax;
    private int colorMin;
    private int colorMax;
    private int speed;
    private int direction;
    private int colorMode;
    private Color[] colors;

    public Mode(String name, int value, int flags, int speedMin, int speedMax, int colorMin, int colorMax, int speed, int direction, int colorMode, Color[] colors) {
        this.name = name;
        this.value = value;
        this.flags = flags;
        this.speedMin = speedMin;
        this.speedMax = speedMax;
        this.colorMin = colorMin;
        this.colorMax = colorMax;
        this.speed = speed;
        this.direction = direction;
        this.colorMode = colorMode;
        this.colors = colors;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getValue() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }

    public int getFlags() {
        return flags;
    }

    public void setFlags(int flags) {
        this.flags = flags;
    }

    public int getSpeedMin() {
        return speedMin;
    }

    public void setSpeedMin(int speedMin) {
        this.speedMin = speedMin;
    }

    public int getSpeedMax() {
        return speedMax;
    }

    public void setSpeedMax(int speedMax) {
        this.speedMax = speedMax;
    }

    public int getColorMin() {
        return colorMin;
    }

    public void setColorMin(int colorMin) {
        this.colorMin = colorMin;
    }

    public int getColorMax() {
        return colorMax;
    }

    public void setColorMax(int colorMax) {
        this.colorMax = colorMax;
    }

    public int getSpeed() {
        return speed;
    }

    public void setSpeed(int speed) {
        this.speed = speed;
    }

    public int getDirection() {
        return direction;
    }

    public void setDirection(int direction) {
        this.direction = direction;
    }

    public int getColorMode() {
        return colorMode;
    }

    public void setColorMode(int colorMode) {
        this.colorMode = colorMode;
    }

    public Color[] getColors() {
        return colors;
    }

    public void setColors(Color[] colors) {
        this.colors = colors;
    }

}

File: openrgb-wrapper/src/main/kotlin/de/lars/openrgbwrapper/models/Mode.kt
package de.lars.openrgbwrapper.models

import com.sun.org.apache.xpath.internal.operations.Mod
import java.nio.ByteBuffer
import java.nio.ByteOrder

import de.lars.openrgbwrapper.utils.BufferUtil.readString
import de.lars.openrgbwrapper.utils.Pair


data class Mode(val name: String, val value: Int, val flags: Int, val speedMin: Int, val speedMax: Int,
                val colorMin: Int, val colorMax: Int, val speed: Int, val direction: Int, val colorMode: Int, val colors: Array<Color>) {

    companion object {

        /**
         * Decode a byte buffer to an array of modes
         */
        fun decode(buffer: ByteArray, bufOffset: Int, numModes: Int): Pair<Array<Mode>, Int> {
            val buf: ByteBuffer = ByteBuffer.wrap(buffer).order(ByteOrder.LITTLE_ENDIAN)
            var offset: Int = bufOffset

            val modes: Array<Mode> = Array(numModes) {
                val namePair: Pair<String, Int> = readString(buf, offset)
                val name = namePair.first
                offset += namePair.second

                val value = buf.getInt(offset)
                offset += 4

                val flags = buf.getInt(offset)
                offset += 4

                val speedMin = buf.getInt(offset)
                offset += 4

                val speedMax = buf.getInt(offset)
                offset += 4

                val colorMin = buf.getInt(offset)
                offset += 4

                val colorMax = buf.getInt(offset)
                offset += 4

                val speed = buf.getInt(offset)
                offset += 4

                val direction = buf.getInt(offset)
                offset += 4

                val colorMode = buf.getInt(offset)
                offset += 4

                val colorCount: Int = buf.getShort(offset).toInt()
                offset += 2
                val colors = Color.decode(buf.array(), offset, colorCount)
                offset += colorCount - 1

                Mode(name, value, flags, speedMin, speedMax, colorMin, colorMax, speed, direction, colorMode, colors)
            }
            return Pair(modes, offset)
        }
    }

}
